File: 01_Kafka_Clients/java/src/main/java/com/masterclass/kafka/OrderConsumer.java
package com.masterclass.kafka;

import io.confluent.kafka.serializers.KafkaAvroDeserializer;
import io.confluent.kafka.serializers.KafkaAvroDeserializerConfig;
import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.kafka.common.serialization.StringDeserializer;

import java.time.Duration;
import java.util.Collections;
import java.util.Properties;

public final class OrderConsumer {
    public static void main(String[] args) {
        int maxMessages = parseMaxMessages(System.getenv("MAX_MESSAGES"));

        Properties props = new Properties();
        props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, KafkaConfig.BOOTSTRAP_SERVERS);
        props.put(ConsumerConfig.GROUP_ID_CONFIG, consumerGroupId());
        props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, autoOffsetReset());
        props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, KafkaAvroDeserializer.class.getName());
        props.put("schema.registry.url", KafkaConfig.SCHEMA_REGISTRY_URL);
        props.put(KafkaAvroDeserializerConfig.SPECIFIC_AVRO_READER_CONFIG, true);

        try (KafkaConsumer<String, OrderEvent> consumer = new KafkaConsumer<>(props)) {
            consumer.subscribe(Collections.singletonList(KafkaConfig.TOPIC));
            System.out.printf("Consuming from '%s'%s ...%n",
                    KafkaConfig.TOPIC,
                    maxMessages > 0 ? " (max " + maxMessages + " messages)" : " (Ctrl+C to stop)");

            Runtime.getRuntime().addShutdownHook(new Thread(consumer::wakeup));

            int consumed = 0;
            while (maxMessages <= 0 || consumed < maxMessages) {
                for (ConsumerRecord<String, OrderEvent> record : consumer.poll(Duration.ofSeconds(1))) {
                    OrderEvent event = record.value();
                    System.out.printf(
                            "Consumed order_id=%s status=%s customer_id=%s amount_cents=%d "
                                    + "partition=%d offset=%d%n",
                            event.getOrderId(),
                            event.getStatus(),
                            event.getCustomerId(),
                            event.getAmountCents(),
                            record.partition(),
                            record.offset());
                    consumed++;
                    if (maxMessages > 0 && consumed >= maxMessages) {
                        System.out.printf("Reached MAX_MESSAGES=%d, exiting.%n", maxMessages);
                        return;
                    }
                }
            }
        }
    }

    private static String consumerGroupId() {
        String value = System.getenv("CONSUMER_GROUP_ID");
        return value == null || value.isBlank() ? "order-fulfillment-java" : value;
    }

    private static String autoOffsetReset() {
        String value = System.getenv("AUTO_OFFSET_RESET");
        return value == null || value.isBlank() ? "earliest" : value;
    }

    private static int parseMaxMessages(String value) {
        if (value == null || value.isBlank()) {
            return 0;
        }
        return Integer.parseInt(value);
    }

    private OrderConsumer() {}
}

File: 01_Kafka_Clients/python/order_consumer.py
#!/usr/bin/env python3
"""Consume order lifecycle events from Kafka."""

from __future__ import annotations

import os
import signal
import sys

from confluent_kafka import DeserializingConsumer
from confluent_kafka.serialization import StringDeserializer
from confluent_kafka.schema_registry import SchemaRegistryClient
from confluent_kafka.schema_registry.avro import AvroDeserializer

from config import BOOTSTRAP_SERVERS, SCHEMA_PATH, SCHEMA_REGISTRY_URL, TOPIC

running = True


def shutdown(_signum, _frame) -> None:
    global running
    running = False


def max_messages() -> int:
    value = os.getenv("MAX_MESSAGES", "").strip()
    return int(value) if value else 0


def build_consumer() -> DeserializingConsumer:
    schema_str = SCHEMA_PATH.read_text(encoding="utf-8")
    registry = SchemaRegistryClient({"url": SCHEMA_REGISTRY_URL})
    avro_deserializer = AvroDeserializer(registry, schema_str)

    return DeserializingConsumer(
        {
            "bootstrap.servers": BOOTSTRAP_SERVERS,
            "group.id": os.getenv("CONSUMER_GROUP_ID", "order-fulfillment-py"),
            "auto.offset.reset": os.getenv("AUTO_OFFSET_RESET", "earliest"),
            "key.deserializer": StringDeserializer("utf_8"),
            "value.deserializer": avro_deserializer,
        }
    )


def main() -> int:
    limit = max_messages()
    signal.signal(signal.SIGINT, shutdown)
    signal.signal(signal.SIGTERM, shutdown)

    consumer = build_consumer()
    consumer.subscribe([TOPIC])
    suffix = f" (max {limit} messages)" if limit > 0 else " (Ctrl+C to stop)"
    print(f"Consuming from '{TOPIC}'{suffix} ...")

    consumed = 0
    try:
        while running and (limit <= 0 or consumed < limit):
            msg = consumer.poll(1.0)
            if msg is None:
                continue
            if msg.error():
                print(f"Consumer error: {msg.error()}", file=sys.stderr)
                continue
            value = msg.value()
            line = (
                f"Consumed order_id={value['order_id']} status={value['status']} "
                f"customer_id={value['customer_id']} amount_cents={value['amount_cents']}"
            )
            if value.get("fulfillment_center"):
                line += f" fulfillment_center={value['fulfillment_center']}"
            print(f"{line} partition={msg.partition()} offset={msg.offset()}")
            consumed += 1
            if limit > 0 and consumed >= limit:
                print(f"Reached MAX_MESSAGES={limit}, exiting.")
                break
    finally:
        consumer.close()

    return 0


if __name__ == "__main__":
    sys.exit(main())
